### ISSUE TITLE
support round robin in pipe iterator

### ISSUE THREAD

**Issue by rudolfix** [dlt-hub/dlt#379]
**Background**
Current iteration strategy exhaust generator steps in pipes in order they were passed to the iterator. This is to have deterministic behavior that you would get from a normal iterator. However there are the cases when we want to iterate with other strategies ie. round robin where we take next item from the next generator, not exhausting the first one first.

Currently the process works as follows:
0. all generator steps in pipes (resources) are added to iterator in reverse.
1. iterator always takes an item from the last generator in the list
2. the items is pushed through the pipeline. it may add more generators to a list
3. item is yielded
4. go to 1

**as a consequence: all the generators that were created by the item will be exhausted (we take last element from the list) before next item form any of the original gens is evaluated**. this guarantees that items are leaving pipeline and this property should be preserved.

**Tasks**
1. * [x] if `gen` yields `None` get next item immediately (default or round robin)
2. * [x] add new config ('next_item_mode`) option to `PipeIteratorConfiguration` that can be `always_last` (current) and `round_robin`
3. * [x]  add option to `extract` method of `Pipeline` to pass that explicitly
4. * [x] it think that `_get_source_item` method is quite good so we provide separate implementation for round robin (method or we subclass)
5. * [x] round robin should only apply to the gens created in step 0 (see above), not to gens that were added as consequence of processing items. in other words even in round robin we still evaluate new gens fully before coming back to old gens
6. * [x] in round robin mode we keep an index of last gen and subtract it modulo

The desired consequence is that if gens return deferred or awaitable items the futures queue fills up to `max_parallel_items` and will execute them in parallel. mind that for sync gens no parallelism will be achieved - just the resources will be evaluated round robin
